doc/developer/design/20240925_network_policies.md

```diff
@@ -0,0 +1,148 @@
+Customers would like to restrict access to Materialize by IP address.
+https://github.com/MaterializeInc/database-issues/issues/4637
+
+## Success Criteria
+- Customers can define a global policy that restricts access to their Materialize environments based on the IP address of the client attempting to connect.
+- Materialize support can unlock an environment where policies prevent access.
+- The console is aware of whether an environment's network policies are blocking a connection it is trying to make.
+- Users can adjust network policies via SQL and in the console.
+
+Nice to haves:
+- Per-role network policies.
+- Policies for sources/sinks.
+- Mitigations to prevent user lockouts; such as ensuring the user is not blocking their current IP.
+- Termination of active connections based on newly applied policies.
+- Policies for egress traffic.
+
+## Out of Scope
+- Preventing out-of-policy traffic from reaching Environmentd.
+ (note this means network policies will not prevent DDOS)
+- Policy inheritance from associated roles. IE if 'bob' is a member of role 'eng'
+we will not apply policies from role 'eng' to 'bob'.
+- Restricting global API access.
+- Restricting access to Frontegg.
+
+## Solution Proposal
+
+#### Overview
+The proposed solution is to use role-based policies with a default network policy that applies to any role without a policy. This can initially be implemented as a global default policy and will be extended to per-user and per-source/sink. The policy will be applied when an attempt is made to establish a new client connection with the coordinator.
+
+#### New Resources
+A new `NetworkPolicy` resource will be added to the catalog.
+```rust
+struct NetworkPolicy {
+    id: NetworkPolicyId
+    name: String,
+    rules: Vec<NetworkPolicyRule>,
+}
+
+enum NetworkPolicyRule {
+    Ingress {
+        action: NetworkPolicyRuleAction,
+        source: IpNet,
+        comment: String
+ }
+}
+
+enum NetworkPolicyRuleAction {
+    Allow
+    // Deny - may be added later
+}
+```
+
+Users will be able to create `NetworkPolicies` directly.  A user must have `CREATENETWORKPOLICY` privileges to create, modify, or destroy network policies. Network policies will be limited to 25 rules. This will be controlled by an LD flag. `NetworkPolicyRules` must be created through a policy. The policy rules implementation will initially only contain an `Allow` variant, but we should be an enum to allow for a `Deny` variant in the future. `NetworkPolicyRule` will be an enum to allow for both ingress and egress policies. Only ingress policies will be initially implemented. `NetworkPolicyRules::Ingress` will also contain a single `IpNet` and a comment text field. Comments have become a standard for rules and greatly increase the manageability and auditability of policies.
+
+Example syntax for creating a network policy
+```sql
+CREATE NETWORK POLICY OFFICE_01 (
+ RULE ( ACTION=ALLOW, SOURCE="10.0.0.0/32", COMMENT="OFFICE IP - 2024-9-28" )
+);
+```
```

Comment:
Overall this looks good to me, I do wonder if we want to separate out `COMMENT` though and instead point folks towards something like:
```
COMMENT ON NETWORK POLICY office_01 is 'OFFICE IP - 2024-9-28'
```
Just so it's the same as other objects. Regardless we should probably pass this through the SQL council!

Reply:
Yes we'd want the comment to be on each rule which is not a resource that's stored outside of the policy. 

... maybe description is a better name for this field.

FWIW my inspiration was, in part, aws security groups

![image](https://github.com/user-attachments/assets/cf4e9f09-37a0-4f50-9230-9d32142abe34)

Reply:
I think comment is good—PostgreSQL (and therefore Materialize) uses "comment" to mean what AWS calls "description" pretty routinely.

Here's my recommendation: force every rule to have a name. Something like this would align with how you declare replicas for unmanaged clusters:

```sql
CREATE NETWORK POLICY OFFICE_01 (
  RULES(
    nikhils_desktop (ACTION=ALLOW, SOURCE="10.0.0.0/32"),
    justins_laptop (ACTION=ALLOW, SOURCE="10.0.0.2/32"),
  )
 );
```

Then you could name a specific rule in the `COMMENT` statement:

```sql
COMMENT ON NETWORK RULE office_01.justins_laptop is 'Laptop assigned to Justin Bradfield on 2024-01-03'
```

It also tees us up to support altering network policies one rule at a time, like:

```
ALTER NETWORK POLICY DROP RULE justins_laptop
```

We don't need to support that first thing, but it's nice to have the option to introduce the granular editing syntax.